src_c/_sdl2/controller.pyx

```diff
@@ -0,0 +1,129 @@
+from . import error
+
+cdef extern from "../pygame.h" nogil:
+    int pgJoystick_Check(object joy)
+    object pgJoystick_New(int);
+    void import_pygame_joystick()
+	
+import_pygame_joystick()
+
+cdef extern from "SDL.h" nogil:
+    void SDL_free(void *mem)
+
+def set_eventstate(state):
+    return SDL_GameControllerEventState(state)
+
+def get_count():
+    """ Returns the number of attached joysticks.
+    """
+    num = SDL_NumJoysticks()
+    if num < 0:
+        raise error()
+    return num
+
+def update():
+    """ Will automatically called by the event loop,
+        not necessary to call this function.
+    """
+    SDL_GameControllerUpdate()
+
+def is_controller(index):
+    """ Check if the given joystick is supported by the game controller interface.
+
+    :param int index: Index of the joystick.
+
+    :return: 1 if supported, 0 if unsupported or invalid index.
+    """
+    return SDL_IsGameController(index)
+
+def name_forindex(index):
+    """ Returns the name of controller,
+        or NULL if there's no name or the index is invalid.
+    """
+    return SDL_GameControllerNameForIndex(index)
+
+cdef class Controller:
+    def __init__(self, int index):
+        """ Create a controller object and open it by given index.
+        
+        :param int index: Index of the joystick.
+        """
+
+        if not SDL_IsGameController(index):
+            raise error('Index is invalid or not a supported joystick.')
+
+        self._controller = SDL_GameControllerOpen(index)
+        self._index = index
+        if not self._controller:
+            raise error('Could not open controller %d.' % index)
+
+    @staticmethod
+    def from_joystick(joy):
+        # https://wiki.libsdl.org/SDL_GameControllerFromInstanceID
+        """ Create a controller object from pygame.joystick.Joystick object.
+        
+        """
+        if not pgJoystick_Check(joy):
+            raise TypeError('should be a pygame.joystick.Joystick object.')
+
+        cdef Controller self = Controller.__new__(Controller)
```

Comment:
Does `Controller(joy.get_id())` work?

Reply:
joy.get_id() means joystick.Joystick.get_id, which is also a number from 0 to joystick.get_count()